### PR TITLE
Implemented viewmodel for RankingWidgetDialog to solve the problem with serialization

### LINKED ISSUE #3615: Collect crashes after minimizing an app when Rank widget popup displayed.
#### Software and hardware versions 
Collect v1.25.1, Android v4.2-10, device used...

#### Problem description
Collect crashes after minimizing/extinguishing an app when Rank widget popup displayed.

#### Steps to reproduce the problem
1. Open Rank widget
2. Click Rank items button
3. Minimize/extinguish the app

#### Expected behavior
The app should not crash. User should be able to restore the app from the background. 

#### Other information 
I have used All widgets form with rank widget.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #3615

#### What has been done to verify that this works as intended?
I tested the fix manually.

#### Why is this the best possible solution? Were any other approaches considered?
It's the solution we discussed in https://github.com/getodk/javarosa/pull/536#issuecomment-633733161

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
It should just fix the issue and have no other impact. We can test just the rank widget.

#### Do we need any specific form for testing your changes? If so, please attach one.
Any form with rank widget.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java [hunk 1]
```diff
@@ -35,57 +36,46 @@
...
-    @Override
-    public void onCreate(Bundle savedInstanceState) {
-        super.onCreate(savedInstanceState);
-        items = (List<SelectChoice>) (savedInstanceState == null
-                ? getArguments().getSerializable(ITEMS)
-                : savedInstanceState.getSerializable(ITEMS));
-        formIndex = (FormIndex) (savedInstanceState == null
-                ? getArguments().getSerializable(FORM_INDEX)
-                : savedInstanceState.getSerializable(FORM_INDEX));
+        viewModel = ViewModelProviders.of(this, new RankingViewModel.Factory(items, formIndex)).get(RankingViewModel.class);
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java]: Since this is deprecated, I think that in new code it makes sense to use the `ViewModelProvider` constructor - `new ViewModelProvider(this, new RankingViewModel.Factory(items, formIndex)).get(RankingViewModel.class)`

File: collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java [hunk 2]
```diff
@@ -35,57 +36,46 @@
...
-        bundle.putSerializable(ITEMS, (Serializable) items);
-        bundle.putSerializable(FORM_INDEX, formIndex);
-        dialog.setArguments(bundle);
+    public RankingWidgetDialog() {
+    }
 
-        return dialog;
+    public RankingWidgetDialog(List<SelectChoice> items, FormIndex formIndex) {
+        this.items = new ArrayList<>(items);
+        this.formIndex = formIndex;
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java:66]: I don't love that these fields still end up existing since the point of the view model is to host them. What I was imagining originally is the activity creating the view model and the dialog fragment accessing it. I wasn't thinking about how the activity in this case is `FormEntryActivity` but currently it's the widget's responsibility to know about the items. 

Given the existing infrastructure, I find this ok for now. It's something @SaumiaSinghal and @seadowg could give another look at when they get to this widget.

**Review by lognaturel — COMMENTED** [review]
It would be nice to have a test but I'm not really seeing a reasonable way to create one. My general sense is that it's worth getting this fix in for now and that the code structure will be reconsidered as part of @SaumiaSinghal's GSoC project anyway. Does that seem ok to you @seadowg, @SaumiaSinghal?

File: collect_app/src/main/java/org/odk/collect/android/widgets/RankingWidget.java [hunk 3]
```diff
@@ -98,7 +98,7 @@ public void onButtonClick(int buttonId) {
         if (formController != null) {
             formController.setIndexWaitingForData(getFormEntryPrompt().getIndex());
         }
-        RankingWidgetDialog rankingWidgetDialog = RankingWidgetDialog.newInstance(savedItems == null ? items : savedItems, getFormEntryPrompt().getIndex());
+        RankingWidgetDialog rankingWidgetDialog = new RankingWidgetDialog(savedItems == null ? items : savedItems, getFormEntryPrompt().getIndex());
         rankingWidgetDialog.show(((FormEntryActivity) getContext()).getSupportFragmentManager(), "RankingDialog");
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/RankingWidget.java:102]: This is not really related to this PR but it just came to mind that I don't really know when `DialotUtils.showIfNotShowing` makes a difference. Should it be used here?

File: collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java:66]: >What I was imagining originally is the activity creating the view model and the dialog fragment accessing it

we can't create viewmodels using activity because then if we have multiple rank widgets they will share the same viewmodel with the same data. We need to create viewmodels using fragment scope.

I didn't think about it much so maybe I'm wrong but generally it would be great if every single widget were a fragment then we could have a viewmodel for every widget easily.

File: collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java]: Fixed.

File: collect_app/src/main/java/org/odk/collect/android/widgets/RankingWidget.java (hunk 3 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/widgets/RankingWidget.java:102]: Can't be sure whether it's needed here I doubt but... generally it's good to use that method it's a defensive approach which could prevent some strange and rare crashes but that method uses bundle to pass data so that's what I need to avoid in this case.

**Review by lognaturel — APPROVED** [review]

**Comment by mmarciniak90** [comment]
**Tested with success**

Verified on Androids: 6.0, 7.0, 8.1, 9.0

Verified cases:
- minimize
- lock screen
- rotation
- reopen from list of last opened apps
- reopen from list of the installed apps
- `Don't keep activities` turned on/off
- rank widgets in field-list

**Comment by kkrawczyk123** [comment]
Verified on Androids: 5.1, 7.1 and 10.

@getodk-bot unlabel "needs testing"
@getodk-bot label "behavior verified"
